src/object/schema_class_truncator.cpp

```diff
@@ -0,0 +1,634 @@
+	return ER_TRUNCATE_PK_REFERRED;
+      }
+
+    /* Find FK-child classes to cascade. */
+    for (fk_ref = pk_constraint->fk_info; fk_ref; fk_ref = fk_ref->next)
+      {
+	if (fk_ref->delete_action == SM_FOREIGN_KEY_CASCADE)
+	  {
+	    if (m_trun_classes.find (fk_ref->self_oid) != m_trun_classes.end ())
+	      {
+		continue;		/* already checked */
+	      }
+
+	    MOP fk_child_mop = ws_mop (&fk_ref->self_oid, NULL);
+	    if (fk_child_mop == NULL)
+	      {
+		assert (er_errid () != NO_ERROR);
+		return er_errid ();
+	      }
+
+	    int partition_type = DB_NOT_PARTITIONED_CLASS;
+	    error = sm_partitioned_class_type (fk_child_mop, &partition_type, NULL, NULL);
+	    if (error != NO_ERROR)
+	      {
+		return error;
+	      }
+
+	    if (partition_type == DB_PARTITION_CLASS)
+	      {
+		/*
+		 * FKs of all partition classes refers to a class that the parittioned class of them referes to.
+		 * But, partition class will be processed when the partitioned class is done.
+		 */
+		continue;
+	      }
+
+	    error = collect_trun_classes (fk_child_mop, is_cascade);
+	    if (error != NO_ERROR)
+	      {
+		return error;
+	      }
+	  }
+	else
+	  {
+	    er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TRUNCATE_CANT_CASCADE, 1, fk_ref->name);
+	    return ER_TRUNCATE_CANT_CASCADE;
+	  }
+      }
+
+    return NO_ERROR;
+  }
+
+  /*
+   * truncate () - truncates classes colleted through collect_trun_classes()
+   *   return: NO_ERROR on success, non-zero for ERROR
+   *
+   *   NOTE: this function truncates several classes which are bond in a partitioning or some foreign keys.
+   *   truncating a class consists of a few steps, each of which is done one by one across all the classess.
+   *   this horizontal processing is necessary that all FKs has to be processed before PKs.
+   */
+  int
+  class_truncator::truncate (const bool is_cascade)
+  {
+    int error = NO_ERROR;
+    std::vector<class_truncate_context> contexts;
+
+    assert (m_class_mop != NULL);
+
+    error = collect_trun_classes (m_class_mop, is_cascade);
+    if (error != NO_ERROR)
+      {
+	return error;
+      }
+
+    try
+      {
+	std::for_each (m_trun_classes.begin(), m_trun_classes.end(), [&contexts] (const OID& oid)
+	{
+	  contexts.emplace_back (oid);
+	});
+      }
+    catch (int &error)
+      {
+	// exception from sm_context constructor
+	return error;
+      }
+
+    /* Save constraints. Or, remove instances from the constraint if impossible. FK first, then non-FK */
+    for (auto &context : contexts)
+      {
+	auto cons_predicate = [] (const SM_CLASS_CONSTRAINT& cons) -> bool
+	{
+	  return cons.type == SM_CONSTRAINT_FOREIGN_KEY;
+	};
+
+	error = context.save_constraints_or_clear (cons_predicate);
+	if (error != NO_ERROR)
+	  {
+	    return error;
+	  }
+
+	/* FK must be dropped earlier than PK, because of referencing */
+	auto saved_cons_predicate = [] (const SM_CONSTRAINT_INFO& cons_info) -> bool
```

Comment:
You're right. It is not necessary, but I thought it is more readable, and otherwise, we have to write this like: 
```
auto saved_cons_predicate = [] (const SM_CONSTRAINT_INFO& cons_info) -> bool
{
    return true;
};
```
However, while considering your review, I've found out that `std::function` can receive `nullptr`. With this, we can make these kinds of functions can be called without a predicate. It looks better to me. What do you think?

Reply:
What I mention is: [cb66b4d](https://github.com/CUBRID/cubrid/pull/2801/commits/cb66b4d9b6520ac81ae6b323a29000f55fad29dc). We can revert it.